src/realm/deppart/byfield_gpu_impl.hpp

```diff
@@ -0,0 +1,261 @@
+      }
+    }
+    // final end offset
+    inst_offsets[field_data.size()] = inst_size;
+
+    RegionInstance inst_entries_instance = this->realm_malloc(inst_size * sizeof(SparsityMapEntry<N,T>), my_mem);
+    SparsityMapEntry<N,T>* d_inst_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(inst_entries_instance, 0).base);
+
+    size_t inst_pos = 0;
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      const IndexSpace<N,T> &inst_space = field_data[i].index_space;
+      if (inst_space.dense()) {
+        // just one rect
+        SparsityMapEntry<N,T> entry;
+        entry.bounds = inst_space.bounds;
+        CUDA_CHECK(cudaMemcpyAsync(d_inst_entries + inst_pos, &entry, sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+        ++inst_pos;
+      } else {
+        auto tmp = inst_space.sparsity.impl()->get_entries();
+        CUDA_CHECK(cudaMemcpyAsync(d_inst_entries + inst_pos, tmp.data(), tmp.size() * sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+        inst_pos += tmp.size();
+      }
+    }
+
+    span<SparsityMapEntry<N,T>> parent_entries;
+    std::vector<SparsityMapEntry<N,T>> parent_entries_vec;
+    RegionInstance parent_entries_instance;
+    SparsityMapEntry<N,T>* d_parent_entries;
+    if (parent_space.dense()) {
+      SparsityMapEntry<N,T> entry;
+      entry.bounds = parent_space.bounds;
+      parent_entries_vec = {entry};
+      parent_entries = span<SparsityMapEntry<N,T>>(parent_entries_vec.data(), 1);
+    } else {
+      parent_entries = parent_space.sparsity.impl()->get_entries();
+    }
+    parent_entries_instance = this->realm_malloc(parent_entries.size() * sizeof(SparsityMapEntry<N,T>), my_mem);
+    d_parent_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(parent_entries_instance, 0).base);
+    CUDA_CHECK(cudaMemcpyAsync(d_parent_entries, parent_entries.data(), parent_entries.size() * sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+    nvtx_range_pop();
+
+    nvtx_range_push("cuda", "Build and intersect input rectangles");
+
+    RegionInstance inst_counters_instance = this->realm_malloc((field_data.size()) * sizeof(uint32_t), my_mem);
+    uint32_t* d_inst_counters = reinterpret_cast<uint32_t*>(AffineAccessor<char,1>(inst_counters_instance, 0).base);
+    CUDA_CHECK(cudaMemsetAsync(d_inst_counters, 0, (field_data.size()) * sizeof(uint32_t), stream), stream);
+
+    RegionInstance inst_offsets_instance = this->realm_malloc((field_data.size() + 1) * sizeof(size_t), my_mem);
+    size_t* d_inst_offsets = reinterpret_cast<size_t*>(AffineAccessor<char,1>(inst_offsets_instance, 0).base);
+    CUDA_CHECK(cudaMemcpyAsync(d_inst_offsets, inst_offsets.data(), (field_data.size() + 1) * sizeof(size_t), cudaMemcpyHostToDevice, stream), stream);
+
+    int flattened_threads = 256;
+    int grid_size = (parent_entries.size() * inst_size + flattened_threads - 1) / flattened_threads;
```

Comment:
this intermediate computation could probably overflow an int32 for a large instance.

Reply:
`int` by default is not necessarily int64_t, so this could still overflow. Also, it is common to extract this logic into a macro or function since it will be used repeatedly.